peps/pep-0767.rst

```diff
@@ -0,0 +1,548 @@
+
+
+    class Member:
+        def __init__(self, id: int) -> None:
+            self.id: ReadOnly[int] = id
+
+...and the protocol described in :ref:`protocols` is now just::
+
+    from typing import Protocol, ReadOnly
+
+
+    class HasName(Protocol):
+        name: ReadOnly[str]
+
+
+    def greet(obj: HasName, /) -> str:
+        return f"Hello, {obj.name}!"
+
+* A subclass of ``Member`` can redefine ``.id`` as a writable attribute or a
+  :term:`descriptor`. It can also :external+typing:term:`narrow` the type.
+* The ``HasName`` protocol succinctly expresses operations available on its attribute.
+* The ``greet`` function can now accept a wide variety of compatible objects,
+  while being explicit about no modifications being done to the input.
+
+
+Specification
+=============
+
+The :external+py3.13:data:`typing.ReadOnly` :external+typing:term:`type qualifier`
+becomes a valid annotation for :term:`attributes <attribute>` of classes and protocols.
+It can be used at class-level or within ``__init__`` to mark individual attributes read-only::
+
+    class Book:
+        id: ReadOnly[int]
+
+        def __init__(self, id: int, name: str) -> None:
+            self.id = id
+            self.name: ReadOnly[str] = name
+
+Type checkers should error on any attempt to reassign or ``del``\ ete an attribute
+annotated with ``ReadOnly``.
+
+Type checkers should also error on any attempt to delete an attribute annotated as ``Final``.
+(This is not currently specified.)
+
+Akin to ``Final`` [#final_mutability]_, ``ReadOnly`` does not influence how
+type checkers perceive the mutability of the assigned object. Immutable :term:`ABCs <abstract base class>`
+and :mod:`containers <collections.abc>` may be used in combination with ``ReadOnly``
+to forbid mutation of such values:
+
+.. code-block:: python
+
+    from collections import abc
+    from dataclasses import dataclass
+    from typing import Protocol, ReadOnly
+
+
+    @dataclass
+    class Game:
+        name: str
+
+
+    class HasGames[T: abc.Collection[Game]](Protocol):
+        games: ReadOnly[T]
+
+
+    def add_games(shelf: HasGames[list[Game]]) -> None:
+        shelf.games.append(Game("Half-Life"))  # ok: list is mutable
+        shelf.games[-1].name = "Black Mesa"    # ok: "name" is not read-only
+        shelf.games = []                       # error: "games" is read-only
+        del shelf.games                        # error: "games" is read-only and cannot be deleted
+
+
+    def read_games(shelf: HasGames[abc.Sequence[Game]]) -> None:
+        shelf.games.append(...)             # error: "Sequence" has no attribute "append"
+        shelf.games[0].name = "Blue Shift"  # ok: "name" is not read-only
+        shelf.games = []                    # error: "games" is read-only
+
+
+All instance attributes of frozen dataclasses and ``NamedTuple`` should be
+implied to be read-only. Type checkers **should not** flag redundant annotations
+of such attributes with ``ReadOnly``:
```

Comment:
I wonder if it's important that type checkers not be allowed to flag this? Personally I would find the redundant `ReadOnly` annotation below to be marginally bad style (particularly that some attributes are explicitly marked and others not) and would welcome a type checker offering an "info" or "warning" message that the `ReadOnly` annotation is redundant here and can be removed. So I might consider specifying instead that a redundant annotation is not an error, but type checkers may optionally choose to warn about it.